src/library/scala/concurrent/Future.scala

```diff
@@ -677,11 +676,20 @@ object Future {
    */
   def firstCompletedOf[T](futures: TraversableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] = {
     val p = Promise[T]()
-    val completeFirst: Try[T] => Unit = p tryComplete _
-    futures foreach { _ onComplete completeFirst }
+    val firstCompleteHandler = new FirstCompleteHandler(p)
+    futures foreach { _ onComplete firstCompleteHandler }
     p.future
   }
 
+  private  class FirstCompleteHandler[T](p : Promise[T]) extends AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
```

Comment:
The addition of this inner class is triggering the binary compatibility error you saw. In bytecode, a private class is actually public, so a Java source file could see it and reference it when compiling against 2.12.4, and then fail to link if run against 2.12.3. (meaning we haven't kept our forwards binary compatibility promise.)

We can either add a whitelist entry, or more simply, define this as an anonymous class inside `firstCompletedOf`

Reply:
since I'm the new kid on the block :sunglasses: I'd appreciate guidance here, which path is preferable? the white list entry or the anonymous class?

I'm also not so sure about the issue introduced by this class, code running with previous versions gets the original impl (lambda), code running with my patch gets the new specialized function impementation... or is it a worry about code compiled against my patch and executed with a previous version of scala? this code would have to make quite an 'effort' to get to this situation as it'd have to either use reflection or Java code in order to access the newly added nested class...